homeassistant/util/hass_dict.py

```diff
@@ -0,0 +1,231 @@
+"""Implementation for HassDict and custom HassKey types."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Generic, TypeVar, assert_type, overload
+
+_T = TypeVar("_T")
+_U = TypeVar("_U")
+
+
+@dataclass(frozen=True)
+class _Key(Generic[_T]):
+    """Base class for Hass key types."""
+
+    name: str
+
+    def __hash__(self) -> int:
+        """Return hash of name."""
+
+        return hash(self.name)
+
+    def __eq__(self, other: Any) -> bool:
+        """Check equality for dict keys to be compatible with str."""
+
+        if isinstance(other, str):
+            return self.name == other
+        if isinstance(other, _Key):
+            return self.name == other.name
+        return False
+
+
+@dataclass(frozen=True, eq=False)
+class HassEntryKey(_Key[_T]):
+    """Key type for integrations with config entries."""
+
+
+@dataclass(frozen=True, eq=False)
+class HassKey(_Key[_T]):
+    """Generic Hass key type."""
+
+
+class HassDict(dict[_Key[Any] | str, Any]):
+    """Custom dict type to provide better value type hints for Hass key types."""
+
+    @overload  # type: ignore[override]
+    def __getitem__(self, key: HassEntryKey[_T], /) -> dict[str, _T]: ...
+
+    @overload
+    def __getitem__(self, key: HassKey[_T], /) -> _T: ...
+
+    @overload
+    def __getitem__(self, key: str, /) -> Any: ...
+
+    def __getitem__(self, key: _Key[Any] | str, /) -> Any:
+        """Forward to super implementation."""
+        return super().__getitem__(key)
+
+    # ------
+    @overload  # type: ignore[override]
+    def __setitem__(self, key: HassEntryKey[_T], value: dict[str, _T], /) -> None: ...
+
+    @overload
+    def __setitem__(self, key: HassKey[_T], value: _T, /) -> None: ...
+
+    @overload
+    def __setitem__(self, key: str, value: Any, /) -> None: ...
+
+    def __setitem__(self, key: _Key[Any] | str, value: Any, /) -> None:
+        """Forward to super implementation."""
+        return super().__setitem__(key, value)
+
+    # ------
+    @overload  # type: ignore[override]
+    def setdefault(
+        self, key: HassEntryKey[_T], default: dict[str, _T], /
+    ) -> dict[str, _T]: ...
+
+    @overload
+    def setdefault(self, key: HassKey[_T], default: _T, /) -> _T: ...
+
+    @overload
+    def setdefault(self, key: str, default: None = None, /) -> Any | None: ...
+
+    @overload
+    def setdefault(self, key: str, default: Any, /) -> Any: ...
+
+    def setdefault(self, key: _Key[Any] | str, default: Any = None, /) -> Any:
+        """Forward to super implementation."""
+        return super().setdefault(key, default)
```

Comment:
If I'm reading this right it means that we end up loosing the speed of the native code implementation of `setdefault` and others because we now have to do dispatch super() in pure-python

Reply:
Last push resolved this concern